sound/soc/sof/loader.c

```diff
@@ -358,6 +360,68 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
 	return ext_man_size;
 }
 
+static int snd_sof_fw_ext_man4_parse(struct snd_sof_dev *sdev, const struct firmware *fw)
+{
+	struct sof_ipc4_fw_modules *module_entry;
+	struct sof_man4_fw_binary_header *fw_header;
+	struct sof_ext_manifest4_hdr *hdr;
+	struct sof_man4_module_config *fm_config;
+	struct sof_man4_module *fm_entry;
+	int i;
+
+	if (fw->size < sizeof(hdr)) {
+		dev_err(sdev->dev, "Invalid fw size\n");
+		return -EINVAL;
+	}
+
+	hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+
+	fw_header = (struct sof_man4_fw_binary_header *)(fw->data + hdr->len +
```

Comment:
"hdr->len" needs to be validated before use. This must not point beyond the actual FW size in any circumstance.

Reply:
_if_ `fw->size == hdr->len + SOF_MAN4_FW_HDR_OFFSET` then there is no space for 'fw_header', no?
Is it mandatory to have 'fw_header'? if it is then the prior check should be extended.